system/Database/BaseBuilder.php

```diff
@@ -2051,6 +2067,72 @@ public function onConstraint($set)
         return $this;
     }
 
+    /**
+     * Sets data source as a query for insert/update/upsert
+     *
+     * @param BaseBuilder|RawSql|string $query
+     */
+    public function fromQuery($query): BaseBuilder
+    {
+        if (! empty($query)) {
```

Comment:
Why skip if `$query = ''`?

Reply:
I guess because it would set the variables and I preferred not to. It would make a call to protectIdentifiers() as well.

```php
if (is_string($query)) {
    $this->QBOptions['fromQuery'] = $query;
    $this->QBKeys                 = $this->db->protectIdentifiers($this->fieldsFromQuery($query));
    $this->QBSet                  = [];
}
```